src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorPullDiagnosticsEndpoint.cs

```diff
@@ -0,0 +1,90 @@
+
+internal class RazorPullDiagnosticsEndpoint : IRazorPullDiagnosticsEndpoint
+{
+    public bool MutatesSolutionState => false;
+
+    public RegistrationExtensionResult GetRegistration(VSInternalClientCapabilities clientCapabilities)
+    {
+        return new RegistrationExtensionResult("_vs_supportsDiagnosticRequests", options: clientCapabilities.SupportsDiagnosticRequests);
+    }
+
+    public TextDocumentIdentifier GetTextDocumentIdentifier(VSInternalDocumentDiagnosticsParams request)
+    {
+        if (request.TextDocument is null)
+        {
+            throw new NotImplementedException("Why would document be null?");
+        }
+
+        return request.TextDocument;
+    }
+
+    public async Task<IEnumerable<VSInternalDiagnosticReport>?> HandleRequestAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext context, CancellationToken cancellationToken)
+    {
+        var result = await GetPullDiagnosticsAsync(request, context, cancellationToken);
+
+        return result;
+    }
+
+    private async Task<IEnumerable<VSInternalDiagnosticReport>?> GetPullDiagnosticsAsync(VSInternalDocumentDiagnosticsParams request, RazorRequestContext requestContext, CancellationToken cancellationToken)
+    {
+        var clientNotifier = requestContext.GetRequiredService<ClientNotifierServiceBase>();
+        var documentContext = requestContext.GetRequiredDocumentContext();
+        var versionedTextDocumentIdentifier = new VersionedTextDocumentIdentifier
+        {
+            Uri = request.TextDocument!.Uri,
+            Version = documentContext.Version,
+        };
+
+        var delegatedParams = new DelegatedDiagnosticParams(versionedTextDocumentIdentifier);
+
+        var result = await clientNotifier.SendRequestAsync<DelegatedDiagnosticParams, VSInternalDiagnosticReport[]?>(
+            RazorLanguageServerCustomMessageTargets.RazorPullDiagnosticEndpointName,
+            delegatedParams,
+            cancellationToken);
+
+        var mappingService = requestContext.GetRequiredService<RazorDocumentMappingService>();
+        var codeDocument = await documentContext.GetCodeDocumentAsync(cancellationToken);
+
+        var mappedResults = MapDelegatedResults(result, codeDocument, mappingService);
+
+        return mappedResults;
+    }
+
+    private IEnumerable<VSInternalDiagnosticReport> MapDelegatedResults(IEnumerable<VSInternalDiagnosticReport>? reports, RazorCodeDocument codeDocument, RazorDocumentMappingService mappingService)
+    {
+        if (reports is null)
+        {
+            return Array.Empty<VSInternalDiagnosticReport>();
+        }
+
+        foreach (var report in reports)
+        {
+            if (report.Diagnostics is not null)
+            {
+                foreach (var diagnostic in report.Diagnostics)
+                {
+                    if (mappingService.TryMapFromProjectedDocumentRange(codeDocument, diagnostic.Range, out var razorRange))
```

Comment:
We only want to map the C# diagnostics.. HTML will be at the correct spot, but the map function doesn't have those smarts.. does it? I don't think it possibly can.

Reply:
This was a good call-out because it made me realize that "Migrate C# & HTML diagnostics" in the issue title had wormed  its way into my brain. Actually the current HtmlCSharpLS endpoint for this does not request HTML diagnostics (and from what I can see none would be provided even if it requested them anyway). Removing that should solve this comment and the one about parallel requests.

Reply:
Wow. So how to HTML diagnostics work?

Reply:
HTML does push 😉

Reply:
I mean.. I suppose I knew that since there are only two options. No idea why I asked! 😛